internal/graph/builder.go

```diff
@@ -0,0 +1,101 @@
+package graph
+
+type CheckResolverBuilder struct {
+	localCheckerOpts       []LocalCheckerOption
+	cacheOpts              []CachedCheckResolverOpt
+	dispatchThrottlingOpts []DispatchThrottlingCheckResolverOpt
+
+	cycleDetectionCheckResolver     *CycleDetectionCheckResolver
+	cachedCheckResolver             *CachedCheckResolver
+	dispatchThrottlingCheckResolver *DispatchThrottlingCheckResolver
+	localCheckResolver              *LocalChecker
+}
+
+// CheckQueryBuilderOpt defines an option that can be used to change the behavior of CheckResolverBuilder
+// instance.
+type CheckQueryBuilderOpt func(checkResolver *CheckResolverBuilder)
+
+// WithLocalCheckerOpts sets the opts to be used to build LocalChecker.
+func WithLocalCheckerOpts(opts ...LocalCheckerOption) CheckQueryBuilderOpt {
+	return func(r *CheckResolverBuilder) {
+		r.localCheckerOpts = opts
+	}
+}
+
+// WithCachedCheckResolverOpts sets the opts to be used to build CachedCheckResolver.
+func WithCachedCheckResolverOpts(opts ...CachedCheckResolverOpt) CheckQueryBuilderOpt {
+	return func(r *CheckResolverBuilder) {
+		r.cacheOpts = opts
+	}
+}
+
+// WithDispatchThrottlingCheckResolverOpts sets the opts to be used to build DispatchThrottlingCheckResolver.
+func WithDispatchThrottlingCheckResolverOpts(opts ...DispatchThrottlingCheckResolverOpt) CheckQueryBuilderOpt {
+	return func(r *CheckResolverBuilder) {
+		r.dispatchThrottlingOpts = opts
+	}
+}
+
+func NewCheckQueryBuilder(opts ...CheckQueryBuilderOpt) *CheckResolverBuilder {
+	checkQueryBuilder := &CheckResolverBuilder{}
+	for _, opt := range opts {
+		opt(checkQueryBuilder)
+	}
+	return checkQueryBuilder
+}
+
+// NewLayeredCheckResolver constructs a CheckResolver that is composed of various CheckResolver layers.
+// Specifically, it constructs a CheckResolver with the following composition:
+//
+//	CycleDetectionCheckResolver  <-----|
+//		CachedCheckResolver              |
+//			DispatchThrottlingCheckResolver |
+//				LocalChecker                   |
+//					CycleDetectionCheckResolver -|
+//
+// The returned CheckResolverCloser should be used to close all resolvers involved in the
+// composition after you are done with the CheckResolver.
+func (c *CheckResolverBuilder) NewLayeredCheckResolver(
+	cacheEnabled bool,
+	dispatchThrottlingEnabled bool,
+) (CheckResolver, CheckResolverCloser) {
```

Comment:
1. I'd rename this to `Build()`
2. Why are we passing in even more flags to this method? `Build()` should accept zero parameters, if we follow the convention from other libraries that also use this pattern.

Reply:
Ideal usage pattern:

```go
import (
  "github.com/openfga/openfga/internal/graph/check"
)

checkResolverBuilder := check.NewCheckResolverBuilder(
    WithCacheEnabled(),
    WithDispatchThrottlingEnabled(),
})

checkResolver := checkResolverBuilder.Build()

checkResolver.ResolveCheck(...)
```

@miparnisari does this align with your ideal outcomes here?

Reply:
Yes